File: solutions/0/0/8/4/p0084_s0.c
/* Solution to Euler Problem 84: Monopoly Odds. */
#include "runner.h"

#define BOARD_SIZE 40
#define STATES (BOARD_SIZE * 3)

static const char *board[BOARD_SIZE] = {
    "GO",   "A1",  "CC1", "A2",   "T1",  "R1",  "B1",  "CH1", "B2",  "B3",
    "JAIL", "C1",  "U1",  "C2",   "C3",  "R2",  "D1",  "CC2", "D2",  "D3",
    "FP",   "E1",  "CH2", "E2",   "E3",  "R3",  "F1",  "F2",  "U2",  "F3",
    "G2J",  "G1",  "G2",  "CC3",  "G3",  "R4",  "CH3", "H1",  "T2",  "H2"
};

static int JAIL;

/* Index of the square whose name matches exactly. */
static int name_index(const char *name) {
    for (int i = 0; i < BOARD_SIZE; i++) {
        if (strcmp(board[i], name) == 0) return i;
    }
    return 0;
}

/* Advance clockwise from `position` to the next square whose name starts with `prefix`. */
static int forward_seek(int position, const char *prefix) {
    int len = (int)strlen(prefix);
    for (int i = 1; i <= BOARD_SIZE; i++) {
        int pos = (position + i) % BOARD_SIZE;
        if (strncmp(board[pos], prefix, len) == 0) return pos;
    }
    return position;
}

/* Distribution over *finishing* squares after stopping on `landing`, written into out[BOARD_SIZE].
   G2J sends to jail; CC/CH draw a card (1/16 each); the Chance "back 3" can land on CC3, so its
   destination is resolved recursively to chain into a Community Chest draw. */
static void resolve(int landing, double out[BOARD_SIZE]) {
    for (int i = 0; i < BOARD_SIZE; i++) out[i] = 0.0;
    const char *name = board[landing];
    if (strcmp(name, "G2J") == 0) {
        out[JAIL] += 1.0;
        return;
    }
    if (strncmp(name, "CC", 2) == 0) {
        /* 2/16 movement cards (Advance to GO, Go to Jail); 14/16 stay. */
        out[0] += 1.0 / 16;
        out[JAIL] += 1.0 / 16;
        out[landing] += 14.0 / 16;
        return;
    }
    if (strncmp(name, "CH", 2) == 0) {
        /* 10/16 movement cards; 6/16 stay. "next R" appears twice (2/16 combined). */
        out[0] += 1.0 / 16;                              /* Advance to GO */
        out[JAIL] += 1.0 / 16;                           /* Go to Jail */
        out[name_index("C1")] += 1.0 / 16;               /* Go to C1 */
        out[name_index("E3")] += 1.0 / 16;               /* Go to E3 */
        out[name_index("H2")] += 1.0 / 16;               /* Go to H2 */
        out[name_index("R1")] += 1.0 / 16;               /* Go to R1 */
        out[forward_seek(landing, "R")] += 1.0 / 16;     /* next Railway */
        out[forward_seek(landing, "R")] += 1.0 / 16;     /* next Railway (again) */
        out[forward_seek(landing, "U")] += 1.0 / 16;     /* next Utility */
        double back[BOARD_SIZE];                          /* Go back 3 squares (may chain via CC3) */
        resolve((landing - 3 + BOARD_SIZE) % BOARD_SIZE, back);
        for (int i = 0; i < BOARD_SIZE; i++) out[i] += back[i] / 16.0;
        out[landing] += 6.0 / 16;
        return;
    }
    out[landing] += 1.0;
}

/* Build the transition matrix over the augmented state square*3 + consecutive_doubles. Tracking the
   doubles count (0/1/2 coming into a turn) lets the third successive double send the player straight
   to jail; any move finishing in jail resets the streak. */
static void build_matrix(int dice_size, double matrix[STATES][STATES]) {
    for (int i = 0; i < STATES; i++)
        for (int j = 0; j < STATES; j++) matrix[i][j] = 0.0;

    double roll_probability = 1.0 / ((double)dice_size * dice_size);
    double fin[BOARD_SIZE];
    for (int position = 0; position < BOARD_SIZE; position++) {
        for (int doubles = 0; doubles < 3; doubles++) {
            int s = position * 3 + doubles;
            for (int first = 1; first <= dice_size; first++) {
                for (int second = 1; second <= dice_size; second++) {
                    if (first == second && doubles == 2) {
                        /* Third consecutive double: go directly to jail, streak resets. */
                        matrix[s][JAIL * 3] += roll_probability;
                        continue;
                    }
                    int next_doubles = (first == second) ? doubles + 1 : 0;
                    int landing = (position + first + second) % BOARD_SIZE;
                    resolve(landing, fin);
                    for (int square = 0; square < BOARD_SIZE; square++) {
                        if (fin[square] == 0.0) continue;
                        int streak = (square == JAIL) ? 0 : next_doubles;
                        matrix[s][square * 3 + streak] += roll_probability * fin[square];
                    }
                }
            }
        }
    }
}

/* Power-iterate the chain to its stationary distribution over the augmented states. */
static void stationary(double matrix[STATES][STATES], double dist[STATES]) {
    for (int i = 0; i < STATES; i++) dist[i] = 1.0 / STATES;
    double nxt[STATES];
    for (int iter = 0; iter < 100000; iter++) {
        for (int j = 0; j < STATES; j++) nxt[j] = 0.0;
        for (int i = 0; i < STATES; i++) {
            double weight = dist[i];
            if (weight == 0.0) continue;
            for (int j = 0; j < STATES; j++) {
                if (matrix[i][j] != 0.0) nxt[j] += weight * matrix[i][j];
            }
        }
        double diff = 0.0;
        for (int j = 0; j < STATES; j++) {
            double d = nxt[j] - dist[j];
            if (d < 0) d = -d;
            if (d > diff) diff = d;
            dist[j] = nxt[j];
        }
        if (diff < 1e-15) break;
    }
}

/* Exact solution via the board's stationary distribution (no simulation, so deterministic): build
   the augmented (square, consecutive doubles) transition matrix, power-iterate to the steady state,
   marginalise to per-square probabilities, and report the three most popular squares as a six-digit
   modal string. */
const char *solve(int argc, char *argv[]) {
    int dice_size = (argc > 1) ? parse_int(argv[1]) : 6;

    JAIL = name_index("JAIL");

    static double matrix[STATES][STATES];
    static double dist[STATES];
    build_matrix(dice_size, matrix);
    stationary(matrix, dist);

    double square_probability[BOARD_SIZE];
    for (int square = 0; square < BOARD_SIZE; square++) {
        square_probability[square] = dist[square * 3] + dist[square * 3 + 1] + dist[square * 3 + 2];
    }

    /* Top-3 by masked linear scans; strict `>` keeps the lowest index on ties. */
    int top[3] = {-1, -1, -1};
    double used[BOARD_SIZE];
    memcpy(used, square_probability, sizeof(square_probability));
    for (int t = 0; t < 3; t++) {
        double best = -1.0;
        int best_idx = 0;
        for (int j = 0; j < BOARD_SIZE; j++) {
            if (used[j] > best) {
                best = used[j];
                best_idx = j;
            }
        }
        top[t] = best_idx;
        used[best_idx] = -1.0;
    }

    char *result = malloc(7);
    if (!result) return NULL;
    snprintf(result, 7, "%02d%02d%02d", top[0], top[1], top[2]);
    return result;
}

File: solutions/0/0/8/4/p0084_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 84: Monopoly Odds [Level 8]. """
from __future__ import annotations

from solver.runners import runner

board: tuple[str, ...] = (
    "GO",
    "A1",
    "CC1",
    "A2",
    "T1",
    "R1",
    "B1",
    "CH1",
    "B2",
    "B3",
    "JAIL",
    "C1",
    "U1",
    "C2",
    "C3",
    "R2",
    "D1",
    "CC2",
    "D2",
    "D3",
    "FP",
    "E1",
    "CH2",
    "E2",
    "E3",
    "R3",
    "F1",
    "F2",
    "U2",
    "F3",
    "G2J",
    "G1",
    "G2",
    "CC3",
    "G3",
    "R4",
    "CH3",
    "H1",
    "T2",
    "H2",
)
board_size: int = len(board)
JAIL: int = board.index("JAIL")


def seek(start: int, prefix: str) -> int:
    """Advance clockwise from ``start`` to the next square whose name begins with ``prefix``."""
    position = start
    while not board[position].startswith(prefix):
        position = (position + 1) % board_size
    return position


def resolve(landing: int) -> dict[int, float]:
    """Distribution over *finishing* squares after stopping on ``landing``.

    Applies the board's square rules: G2J sends the player to jail; CC and CH draw a card
    (a uniform 1/16 each, the rest leaving the player in place). The Chance "go back 3" can
    land on CC3, so its destination is resolved recursively to chain into a Community Chest draw.
    """
    name = board[landing]
    if name == "G2J":
        return {JAIL: 1.0}
    if name.startswith("CC"):
        # 2/16 movement cards (Advance to GO, Go to Jail); 14/16 stay.
        return {0: 1 / 16, JAIL: 1 / 16, landing: 14 / 16}
    if name.startswith("CH"):
        # 10/16 movement cards; 6/16 stay. "next R" appears twice (2/16 combined).
        targets = [
            0,                          # Advance to GO
            JAIL,                       # Go to Jail
            board.index("C1"),          # Go to C1
            board.index("E3"),          # Go to E3
            board.index("H2"),          # Go to H2
            board.index("R1"),          # Go to R1
            seek(landing + 1, "R"),     # next Railway
            seek(landing + 1, "R"),     # next Railway (again)
            seek(landing + 1, "U"),     # next Utility
        ]
        distribution: dict[int, float] = {}
        for target in targets:
            distribution[target] = distribution.get(target, 0.0) + 1 / 16
        # "Go back 3 squares" may land on CC3, which then triggers a Community Chest draw.
        for square, probability in resolve((landing - 3) % board_size).items():
            distribution[square] = distribution.get(square, 0.0) + probability / 16
        distribution[landing] = distribution.get(landing, 0.0) + 6 / 16
        return distribution
    return {landing: 1.0}


def transition_matrix(dice_size: int) -> list[list[float]]:
    """Build the transition matrix over the augmented state ``square * 3 + consecutive_doubles``.

    Tracking the consecutive-doubles count (0, 1 or 2 coming into a turn) lets the third
    successive double send the player straight to jail. Any move that finishes in jail resets
    the streak. State count is ``board_size * 3``; an entry is the probability of finishing in
    one state given the player started a turn in another.
    """
    states = board_size * 3
    matrix = [[0.0] * states for _ in range(states)]
    roll_probability = 1 / (dice_size * dice_size)
    for position in range(board_size):
        for doubles in range(3):
            row = matrix[position * 3 + doubles]
            for first in range(1, dice_size + 1):
                for second in range(1, dice_size + 1):
                    if first == second and doubles == 2:
                        # Third consecutive double: go directly to jail, streak resets.
                        row[JAIL * 3] += roll_probability
                        continue
                    next_doubles = doubles + 1 if first == second else 0
                    landing = (position + first + second) % board_size
                    for square, probability in resolve(landing).items():
                        streak = 0 if square == JAIL else next_doubles
                        row[square * 3 + streak] += roll_probability * probability
    return matrix


def stationary_distribution(matrix: list[list[float]]) -> list[float]:
    """Power-iterate the chain to its stationary distribution over the augmented states."""
    states = len(matrix)
    distribution = [1 / states] * states
    for _ in range(100_000):
        nxt = [0.0] * states
        for i, weight in enumerate(distribution):
            if weight:
                row = matrix[i]
                for j, probability in enumerate(row):
                    if probability:
                        nxt[j] += weight * probability
        if max(abs(a - b) for a, b in zip(nxt, distribution)) < 1e-15:
            return nxt
        distribution = nxt
    return distribution


@runner.main
def solve(*args: str) -> str:
    """Exact solution via the board's stationary distribution (no simulation, so the result is
    deterministic). Build the transition matrix over the augmented state (square, consecutive
    doubles), power-iterate to the steady state, marginalise to per-square probabilities, and
    report the three most popular squares as a six-digit modal string."""
    dice_size = runner.parse_int(args[0])

    distribution = stationary_distribution(transition_matrix(dice_size))
    square_probability = [sum(distribution[square * 3:square * 3 + 3]) for square in range(board_size)]
    ranked = sorted(range(board_size), key=lambda square: (-square_probability[square], square))
    return "".join(f"{square:02d}" for square in ranked[:3])


if __name__ == "__main__":
    raise SystemExit(solve())
